robotoff/insights/importer.py

```diff
@@ -705,85 +772,134 @@ def import_product_predictions(
+            if prediction_types >= required_prediction_types:
+                selected_barcodes.append(barcode)
+
+            if selected_barcodes:
+                predictions = [
+                    Prediction(**p)
+                    for p in get_product_predictions(
+                        selected_barcodes, list(required_prediction_types)
+                    )
+                ]
+                imported += importer.import_insights(
+                    predictions, server_domain, automatic, product_store
+                )
+    return imported
+
+
+def import_predictions(
+    product_predictions: Iterable[ProductPredictions],
+    product_store: DBProductStore,
+    server_domain: str,
+) -> Dict[str, Set[PredictionType]]:
+    """Check validity and import provided ProductPredictions."""
     product_predictions = [
-        p for p in product_predictions if is_valid_product_predictions(p, product_store)
+        p
+        for p in product_predictions
+        if is_valid_product_predictions(p, product_store[p.barcode])
     ]
 
     predictions_imported = 0
-    for barcode, product_predictions_group in itertools.groupby(
+    updated_prediction_types_by_barcode: Dict[str, Set[PredictionType]] = {}
+    for barcode, product_predictions_iter in itertools.groupby(
         sorted(product_predictions, key=operator.attrgetter("barcode")),
         operator.attrgetter("barcode"),
     ):
+        product_predictions_group = list(product_predictions_iter)
         predictions_imported += import_product_predictions(
             barcode, product_predictions_group, server_domain
         )
+        updated_prediction_types_by_barcode.setdefault(barcode, set())
+        updated_prediction_types_by_barcode[barcode] |= set(
+            itertools.chain.from_iterable(
+                (prediction.type for prediction in x.predictions)
+                for x in product_predictions_group
+            )
+        )
     logger.info(f"{predictions_imported} predictions imported")
+    return updated_prediction_types_by_barcode
 
-    prediction_type: PredictionType
-    insight_group: Iterable[ProductPredictions]
-    imported = 0
 
-    for prediction_type, insight_group in itertools.groupby(
-        sorted(product_predictions, key=operator.attrgetter("type")),
-        operator.attrgetter("type"),
-    ):
-        insight_type = InsightType[prediction_type]
-        if insight_type not in importers:
-            importers[insight_type] = InsightImporterFactory.create(
-                insight_type, product_store
-            )
-        importer = importers[insight_type]
-        if importer is not None:
+def refresh_insights(
+    barcode: str,
+    server_domain: str,
+    automatic: bool,
+    product_store: Optional[DBProductStore] = None,
+) -> int:
+    """Refresh all insights for specific product.
+
+    All predictions are fetched, and insights are created/deleted by each
+    InsightImporter.
```

Comment:
At first, I didn't realise the difference with import_insights (because of code pattern), I propose to be more explicit:

```suggestion
    InsightImporter.
    
    .. note:: this is different from `import_insights`, 
       because here, there is no prediction creation. 
       It's just an refresh based on actual database predictions.
```

Reply:
Although, I'm a bit lost on why we have those two modes of operations.

Reply:
I've added this to make it clearer:
```
    This is different from `import_insights`, because here, there is no
    prediction creation.  It's just an refresh based on current database
    predictions. It's useful to refresh insights after an Product Opener
    update (some insights may be invalid).
```